integration/test/Testlib/Cannon.hs

```diff
@@ -293,6 +326,74 @@ awaitNMatchesResult nExpected tSecs checkMatch ws = go nExpected [] []
               }
     refill = mapM_ (liftIO . atomically . writeTChan (wsChan ws))
 
+awaitAtLeastNMatchesResult ::
+  HasCallStack =>
+  -- | Minimum number of matches
+  Int ->
+  -- | Timeout in seconds
+  Int ->
+  -- | Selection function. Exceptions are *not* caught.
+  (Value -> App Bool) ->
+  WebSocket ->
+  App AwaitAtLeastResult
+awaitAtLeastNMatchesResult nExpected tSecs checkMatch ws = go 0 [] []
+  where
+    go nSeen nonMatches matches = do
+      mEvent <- awaitAnyEvent tSecs ws
+      case mEvent of
+        Just event ->
+          do
+            isMatch <- checkMatch event
+            if isMatch
+              then go (nSeen + 1) nonMatches (event : matches)
+              else go nSeen (event : nonMatches) matches
+        Nothing -> do
+          refill nonMatches
+          pure $
+            AwaitAtLeastResult
+              { success = nSeen >= nExpected,
+                nMatchesExpectedMin = nExpected,
+                nMatchesExpectedMax = Nothing,
+                matches = reverse matches,
+                nonMatches = reverse nonMatches
+              }
+    refill = mapM_ (liftIO . atomically . writeTChan (wsChan ws))
+
+awaitNToMMatchesResult ::
```

Comment:
To help with the flaky tests. We want to ensure that they run at least a minimum number of times, but not too many times either. This should hold us over until we can properly address why notifications are being sent more than expected.